### PR TITLE
(PUP-10524) Address spec failures when running with facter 4

### PR DESCRIPTION
**Pull request by melissa**
Fixes still required
- [x] $CHILD_STATUS returns nil
- [x] trace logging

### CODE REVIEW

**Comment by IrimieBogdan** [comment]
@melissa this is great!

The trace problem was fixed today in https://github.com/puppetlabs/facter-ng/pull/527/files

The test that fail with the `$CHILD_STATUS` error mock `texecute` or `execute` from `execution.rb`. Since the execution is mocked, a new process is never created and there is no exit status. It was just a coincidence that Facter 2 spawned a new process and it's exit status was interpreted as the exit status from `texecute` or `execute` 

In order to fix the test we can add a before block to the test e.g
```
  before(:all) do
    `(exit 0)`
  end
```

With the changes from this branch, the `$CHILD_STATUS` fix and Facter 4 master branch, the unit test pass
```
Finished in 9 minutes 53 seconds (files took 8.71 seconds to load)
25800 examples, 0 failures, 39 pending
```

**Comment by IrimieBogdan** [comment]
@melissa I pushed the changes for `$CHILD_STATUS` on your branch and released Facter 4.0.23 today (with the trace fix). Puppet test should pass with those changes.

File: spec/unit/provider/service/redhat_spec.rb [hunk 1]
```diff
@@ -4,6 +4,10 @@
          if: Puppet.features.posix? && !Puppet::Util::Platform.jruby?do
   let(:provider_class) { Puppet::Type.type(:service).provider(:redhat) }
 
+  before(:all) do
```
Comment (melissa) [inline spec/unit/provider/service/redhat_spec.rb:7]: Thank you @IrimieBogdan ! Could you add a comment to the code here explaining why we have to do this? I can see someone getting really confused a year from now as to why this is needed :)

Comment (IrimieBogdan) [reply spec/unit/provider/service/redhat_spec.rb:12]: Added a comments describing the fix and why it is required.

**Pull request merged**